pkg/distributor/tenant_topic_tee.go

```diff
@@ -0,0 +1,395 @@
+func (r *ShardedPartitionResolver) topicName(tenant string, shard int32) string {
+	return fmt.Sprintf("%s.%s.%d", r.topicPrefix, tenant, shard)
+}
+
+// TenantTopicWriter implements the Tee interface by writing logs to Kafka topics
+// based on tenant IDs. Each tenant's logs are written to a dedicated topic.
+type TenantTopicWriter struct {
+	cfg      TenantTopicConfig
+	producer *client.Producer
+	logger   log.Logger
+	resolver PartitionResolver
+	limits   IngestionRateLimits
+
+	// Metrics
+	teeBatchSize    prometheus.Histogram
+	teeQueueLatency prometheus.Histogram
+	teeErrors       *prometheus.CounterVec
+}
+
+// NewTenantTopicWriter creates a new TenantTopicWriter
+func NewTenantTopicWriter(
+	cfg TenantTopicConfig,
+	kafkaClient *kgo.Client,
+	limits IngestionRateLimits,
+	reg prometheus.Registerer,
+	logger log.Logger,
+) (*TenantTopicWriter, error) {
+	strategy, err := ParseStrategy(cfg.Strategy)
+	if err != nil {
+		return nil, fmt.Errorf("invalid strategy: %w", err)
+	}
+
+	var resolver PartitionResolver
+	switch strategy {
+	case SimpleStrategy:
+		resolver = NewSimplePartitionResolver(cfg.TopicPrefix)
+	case AutomaticStrategy:
+		resolver = NewShardedPartitionResolver(kafkaClient, cfg.TopicPrefix)
+	default:
+		return nil, fmt.Errorf("unknown strategy %q", strategy.String())
+	}
+
+	t := &TenantTopicWriter{
+		cfg:    cfg,
+		logger: logger,
+		limits: limits,
+		producer: client.NewProducer(
+			kafkaClient,
+			int64(cfg.MaxBufferedBytes),
+			reg,
+		),
+		resolver: resolver,
+		teeBatchSize: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    "loki_distributor_tenant_topic_tee_batch_size_bytes",
+			Help:    "Size in bytes of batches sent to Kafka by the tenant topic tee",
+			Buckets: prometheus.ExponentialBucketsRange(1<<10, 100<<20, 10),
+		}),
+		teeQueueLatency: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    "loki_distributor_tenant_topic_tee_queue_duration_seconds",
+			Help:    "Duration in seconds spent waiting in queue by the tenant topic tee",
+			Buckets: prometheus.DefBuckets,
+		}),
+		teeErrors: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "loki_distributor_tenant_topic_tee_errors_total",
+			Help: "Total number of errors encountered by the tenant topic tee",
+		}, []string{"reason"}),
+	}
+
+	if reg != nil {
+		reg.MustRegister(t.teeBatchSize, t.teeQueueLatency, t.teeErrors)
+	}
+
+	return t, nil
+}
+
+func (t *TenantTopicWriter) partitionsForRateLimit(bytesRateLimit float64) uint32 {
+	targetThroughputPerPartition := float64(10 << 20) // 10 MB
+
+	target := uint32(math.Round(bytesRateLimit / targetThroughputPerPartition))
+
+	// Calculate partitions, ensuring at least 1
+	return max(target, 1)
+}
+
+// Duplicate implements the Tee interface
+func (t *TenantTopicWriter) Duplicate(tenant string, streams []KeyedStream) {
+	go t.write(tenant, streams)
+}
+
+func (t *TenantTopicWriter) write(tenant string, streams []KeyedStream) {
+	if len(streams) == 0 {
+		return
+	}
```

Comment:
tiniest nit: this condition could be moved to TenantTopicWrite.Duplicate to avoid spawning a goroutine when there's no work to do